### ISSUE TITLE
False positive for CA1067

### ISSUE THREAD

**Issue by manfred-brands** [dotnet/roslyn-analyzers#3983]
### Describe the bug

Related to #2027 
Base class implements `Equals(object)` and `IEquatable<BaseClass>`. 
Derived class overrides `IEquatable<BaseClass>`. It should not have to override `Equals(object)`.

### Steps To Reproduce

    public abstract class BaseClass : IEquatable<BaseClass>
    {
        public abstract bool Equals([AllowNull] BaseClass other);

        public sealed override bool Equals(object obj)
        {
            return Equals(obj as BaseClass);
        }
    }

    public class DerivedClass : BaseClass, IEquatable<DerivedClass>
    {
        public override bool Equals([AllowNull] BaseClass other)
        {
            return Equals(other as DerivedClass);
        }

        public bool Equals([AllowNull] DerivedClass other)
        {
            return true;
        }
    }

    internal sealed class TestEquals
    {
        [Test]
        public void EqualsObject()
        {
            var instance1 = new DerivedClass();
            var instance2 = new DerivedClass();

            Assert.That(Equals(instance1, instance2), Is.True);
        }
    }

### Expected behavior

CA1067 is not raised for `DerivedClass`

### Actual behavior

CA1067 is raised for `DerivedClass`

### Analyzer

**Package**: [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

**Version**: v3.0.0 (Latest)

**Diagnostic ID**: [CA1067](https://docs.microsoft.com/visualstudio/code-quality/ca1067)

**Comment by Evangelink** [comment]
Seems to make sense to me.

**Comment by manfred-brands** [comment]
I see if I can change this in the EquatableAnalyzer.

**Comment by Evangelink** [comment]
I think we need to be careful about the condition on which we stop reporting the rule. In your case, the design is good so `Equals(object)` calls `Equals(Base)` which is abstract (note it would be ok with `virtual` too) but if you have something a little different then you end-up with broken behavior. 

Note that while writing this, I realize that we shall have another analyzer ensuring that the various overloads of `Equals` are behaving correctly (i.e. just all wrapping down to one implementation) and maybe that's ok to simply relax this rule with not much constraint.

**Comment by manfred-brands** [comment]
This rule only checks *if* there is an override, not whether it is a good one.
I think it comes down to a class always needs to override its base class `Equals`, either `Equals(object)` or `Equals(base)`.

Implemented this and added two tests, one following my example and one where the derived class does not override `Equals(Base)`

**Comment by mavasani** [comment]
Alright, so I think I'd like to get opinions on few more folks here - tagging @sharwell @jmarolf @stephentoub 

IMO, the pattern described in the issue where an abstract type seals off `bool Equals(object)` seems very weird to me, and certainly not something I have seen often. The more common and recommended pattern would be that `bool Equals(object)` will be an unsealed override for unsealed types, and CA1067 will help ensure the sub-types will override this method, invoking `base.Equals` to ensure base equality and also performing equality checks for sub-type specific fields. If a user is indeed implementing the pattern as suggested in the issue description, then a CA1067 suppression with appropriate comments acts as good documentation for violating this pattern.

I would like to propose this issue be resolved as by-design.

Also tagging @gewarren as an FYI. If this issue is indeed not accepted, dotnet/docs repo should revert the doc that was already added for [CA1071](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1071) as per the open PR for this issue https://github.com/dotnet/roslyn-analyzers/pull/3992. Going forward, we should ensure that doc PRs for new rules are not merged prior to either the analyzer PR being merged AND the underlying analyzer issue being tagged with `Approved-Rule` label by someone on the team.

**Comment by mavasani** [comment]
NOTE: I am fine with tightening the CA1067 to bail out on type chains where unsealed type seal `bool Equal(object)` to avoid the false positive assuming user is doing something wonky. However, this pattern does not seem to warrant a new analyzer rule CA1071 as proposed in https://github.com/dotnet/roslyn-analyzers/pull/3992 to detect and verify such patterns.

**Comment by manfred-brands** [comment]
The sealing of the method is irrelevant and only done so derived classes cannot add a useless override.
The concept is that every derived class has to override the immediate base class `bool Equals(BaseClass)` not always `bool Equals(object)`.

Let me try a different example:
```
internal class Fruit : IEquatable<Fruit>
{
    protected Fruit(Color color) => Color = color;

    public Color Color { get; }

    public virtual bool Equals(Fruit? other) => other?.Color == Color;

    public override bool Equals(object? obj) => Equals(obj as Fruit);

    public override int GetHashCode() => Color.GetHashCode();
}
```

I then have two derived classes: Banana and Lemon both Yellow.
Both implementing the suggested override `bool Equals(object)` method

```
internal class Banana : Fruit, IEquatable<Banana>
{
    public Banana() : base(Color.Yellow)
    {  }

    public bool Equals(Banana? other) => base.Equals(other);

    public override bool Equals(object? obj) => Equals(obj as Banana);

    public override int GetHashCode() => base.GetHashCode();
}

internal class Lemon : Fruit, IEquatable<Lemon>
{
    public Lemon() : base(Color.Yellow)
    {  }

    public bool Equals(Lemon? other) => base.Equals(other);

    public override bool Equals(object? obj) => Equals(obj as Lemon);

    public override int GetHashCode() => base.GetHashCode();
}
```

But this is flawed:

```
public void TestEquals()
{
    Banana aBanana = new Banana();
    Lemon aLemon = new Lemon();

    Assert.That(Equals(aBanana, aLemon), Is.False, "Object.Equals");
    Assert.That(aBanana.Equals(aLemon), Is.False, "IEquatable");
}
```

The first test succeeds as it calls the suggested overload, the 2nd fails as it calls the `bool Equals(Fruit)` method which only checks for color.

Changing the derived classes to override the base class Equals, fixes both assertions:

```
internal class Banana : Fruit, IEquatable<Banana>
{
    public Banana() : base(Color.Yellow)
    { }

    public bool Equals(Banana? other) => base.Equals(other);

    public override bool Equals(Fruit? other) => Equals(other as Banana);
}

internal class Lemon : Fruit, IEquatable<Lemon>
{
    public Lemon() : base(Color.Yellow)
    { }

    public bool Equals(Lemon? other) => base.Equals(other);

    public override bool Equals(Fruit? other) => Equals(other as Lemon);
}
```

I don't mind dropping CA1071 and replace the wording/implementation of CA1067 to check for base class override.

**Comment by Evangelink** [comment]
Equality in the context of inheritance is always a bit of a nightmare...

Reading again at all thread, I don't know if I wouldn't be more inclined to tweak the rule to force having all overloads of `Equals` on a non-sealed type `virtual`/`abstract` and to force any sub-type adding any field to override the various equality overloads. So in the case of the `Fruit`, `Banana` and `Lemon` would have to implement `Equals(object)`, `Equals(Banana/Lemon)` and `Equals(Fruit)`.

Whichever the direction we go to, I think it'd be nice to update the guidelines and to make sure the analyzer(s) help with going on the recommended direction.

**Comment by sharwell** [comment]
As best I can tell, the derived type should not be adding `IEquatable<DerivedClass>`. Does this analyzer still report a warning for this case?

```csharp
public abstract class BaseClass : IEquatable<BaseClass>
{
    public abstract bool Equals([AllowNull] BaseClass other);

    public sealed override bool Equals(object obj)
    {
        return Equals(obj as BaseClass);
    }
}

public class DerivedClass : BaseClass
{
    public override bool Equals([AllowNull] BaseClass other)
    {
        return true;
    }
}

internal sealed class TestEquals
{
    [Test]
    public void EqualsObject()
    {
        var instance1 = new DerivedClass();
        var instance2 = new DerivedClass();

        Assert.That(Equals(instance1, instance2), Is.True);
    }
}
```

**Comment by Evangelink** [comment]
No complain with `"Microsoft.CodeAnalysis.NetAnalyzers" Version="5.0.1"`.

**Comment by sharwell** [comment]
Based on https://github.com/dotnet/roslyn-analyzers/issues/3983#issuecomment-749663385, it appears that this is likely to be an exceptionally rare edge case. I'm fine with either of the following:

* Do nothing
* Suppress CA1067 for the case where a derived type implements `IEquatable<T2>` _and_ inherits an implementation of `IEquatable<T1>` _and_ a base type has sealed `Equals(object)`

The conditions for the second one could easily be deemed so narrow that this is better suited for a project-specific diagnostic suppressor than a potentially-costly exception in the rule itself.

**Comment by mavasani** [comment]
Lets go with 1. `Do nothing` for now and wait for more feedback/customer request before doing the extra code churn for a rare scenario.
